lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1637,6 +1638,16 @@ class HSAQueue final : public KalmarQueue
             DBOUT(DB_CMD2, "No future found in wait, enqueued marker into " << *this << "\n");
         }
         back()->getFuture()->wait();
+        if (HCC_FLUSH_ON_WAIT) {
+            // aggressively cleanup resources
+            // but keep back() as a valid HSAOp, so decrement current insert index twice
+            int back_op_index = decrement(asyncOpsIndex); // index of back() op
+            int index = decrement(back_op_index); // index of first op to free
+            while (asyncOps[index] != nullptr && index != back_op_index) {
```

Comment:
Why is the asyncOps[index] != nullptr check needed in the while condition?

Reply:
When the HSAQueue is created, its `std::vector< std::shared_ptr<HSAOp> > asyncOps` is default constructed with size `HCC_ASYNCOPS_SIZE`.  This should mean they are all initialized to `nullptr`.  The `asyncOps` vector is treated as a circular buffer via the `asyncOpsIndex` for the current insert index.  The flushing code starts at the `index` prior to the youngest op (`back_op_index`).   This `index` may refer to a nullptr instead of a valid HSAOp because we assign nullptr to HSAOps to trigger their ref-counted destructors.  The check will help short-circuit the while loop once we hit the first non-existent (`nullptr`) HSAOp in the circular buffer.  Otherwise, we would traverse the entire `HCC_ASYNCOPS_SIZE` items, currently 16k in size, for every `wait()`.

Reply:
If we want to avoid traversing large portions of the circular buffer, perhaps using the new environment variable as an upper bound count of items to reclaim could ensure that?

Reply:
Partially cleaning up the async op vector would break the dependence chain.  It "may" still work but undesirable.  The threshold doesn't efficiently solve the problem that the short-circuit in the while loop could do here.

Reply:
It's kind of counter-intuitive to nullify a non-null pointer (even though it is just decrementing ref count). If non-null asyncOps pointer can be reclaimed, does that mean the getPendingAsyncOps() logic needs to be updated?

Reply:
getPendingAsyncOps is still correct but just not optimized, it's a function for debugging anyway